scripts/run_distributed_dataloaders.py

```diff
@@ -0,0 +1,412 @@
+                for _ in range(world_size)]
+
+    # Add explicit barrier before all_gather
+    logger.info(f"Rank {rank}: Before barrier")
+    dist.barrier()
+    logger.info(f"Rank {rank}: After barrier, before all_gather")
+
+    dist.all_gather(gathered, tensor)
+    gathered = [t.item() for t in gathered]
+
+    logger.info(f"Rank {rank}: Can see processes {gathered}")
+    return rank, local_rank, world_size
+
+
+def get_dataset_paths(rank, datasets_per_rank, all_paths):
+    """
+    Distribute dataset paths across ranks.
+    Each rank gets datasets_per_rank consecutive paths.
+    """
+    start_idx = rank * datasets_per_rank
+    end_idx = min(start_idx + datasets_per_rank, len(all_paths))
+
+    # If we run out of datasets, wrap around
+    if start_idx >= len(all_paths):
+        start_idx = start_idx % len(all_paths)
+        end_idx = min(start_idx + datasets_per_rank, len(all_paths))
+
+    paths = all_paths[start_idx:end_idx]
+
+    # wrap around when there are more ranks than datasets
+    if len(paths) < datasets_per_rank:
+        additional_needed = datasets_per_rank - len(paths)
+        paths = np.concatenate([paths, all_paths[:additional_needed]])
+
+    return paths
+
+
+def gather_dataset_info(dataloader, rank, world_size):
+    """
+    Gather dataset keys and neuron counts from all distributed dataloaders.
+
+    Args:
+        dataloader: The local dataloader
+        rank: Process rank
+        world_size: Total number of processes
+
+    Returns:
+        dict: Combined dictionary with all keys and their neuron counts
+    """
+    # Get local dataset info
+    local_info = {}
+    for k, v in dataloader.loaders.items():
+        try:
+            batch = next(iter(v))
+            n_neurons = batch["responses"].shape[-1]
+            local_info[k] = n_neurons
+        except StopIteration:
+            # Handle empty dataloader case
+            logger.warning(f"Rank {rank}: Empty dataloader for key {k}")
+            continue
+
+    # Convert local info to a serializable format
+    # [(key1, neurons1), (key2, neurons2), ...]
+    local_items = list(local_info.items())
+
+    # Determine maximum number of items across all ranks
+    item_counts = [torch.tensor([len(local_items)], device="cuda") for _ in range(world_size)]
+    dist.all_gather(item_counts, torch.tensor([len(local_items)], device="cuda"))
+    max_items = max([count.item() for count in item_counts])
+
+    # Pad local items list if needed
+    if len(local_items) < max_items:
+        # Use a special padding value that we can filter out later
+        local_items.extend([("__pad__", -1) for _ in range(max_items - len(local_items))])
+
+    # For each item, gather its key and neuron count
+    all_keys = []
+    all_neurons = []
+
+    for i in range(max_items):
+        if i < len(local_items):
+            key, neurons = local_items[i]
+            key_tensor = torch.tensor([ord(c) for c in key] + [0] * (50 - len(key)), device="cuda")
```

Comment:
what is `50` here and in line 148. Could we me make it a variable with some type of a meaningful name?